In [ ]:
# Together with all the necessary packages, use the following code to upload the us crime data set.
import pandas as pd
# Please note that this is a text data set not a csv file.
uc = pd.read_csv('uscrime.txt', delimiter ='\t')

More details about the data is here. http://www.statsci.org/data/general/uscrime.html

# Problem 1 [20]
This data has 16 variables. Crime is the dependent variable. Starting with all 15 independent variables, find a multiple linear regression model that has only six significant variables ( p-value $<0.1$) while keeping $R^2\geq 75\%$. For this, you will have to remove one independent variable at a time from the original model, create a new model with the ramaining variables and stop when all the variables have $p$-value $<0.10$. Report the summary of the final model together with the coefficents. Comment on the final model.

In [6]:
import pandas as pd
import numpy as np
import matplotlib as mpt
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm
uc = pd.read_csv('uscrime.txt', delimiter ='\t')

In [7]:
uc.columns

Index(['M', 'So', 'Ed', 'Po1', 'Po2', 'LF', 'M.F', 'Pop', 'NW', 'U1', 'U2',
       'Wealth', 'Ineq', 'Prob', 'Time', 'Crime'],
      dtype='object')

In [8]:
y = uc['Crime']
x = uc[['M', 'So', 'Ed', 'Po1', 'Po2', 'LF', 'M.F', 'Pop', 'NW', 'U1', 'U2',
       'Wealth', 'Ineq', 'Prob', 'Time']]
x = sm.add_constant(x)

In [9]:
model = sm.OLS(y,x)
result = model.fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Crime   R-squared:                       0.803
Model:                            OLS   Adj. R-squared:                  0.708
Method:                 Least Squares   F-statistic:                     8.429
Date:                Thu, 14 Oct 2021   Prob (F-statistic):           3.54e-07
Time:                        15:54:39   Log-Likelihood:                -308.01
No. Observations:                  47   AIC:                             648.0
Df Residuals:                      31   BIC:                             677.6
Df Model:                          15                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -5984.2876   1628.318     -3.675      0.001   -9305.265   -2663.310
M             87.8302     41.714      2.106      0.043       2.754     172.906
So            -3.8035    148.755     -0.026      0.980    -307.192     299.585
Ed           188.3243     62.088      3.033      0.005      61.694     314.954
Po1          192.8043    106.110      1.817      0.079     -23.608     409.216
Po2         -109.4219    117.478     -0.931      0.359    -349.019     130.175
LF          -663.8261   1469.729     -0.452      0.655   -3661.358    2333.706
M.F           17.4069     20.354      0.855      0.399     -24.105      58.919
Pop           -0.7330      1.290     -0.568      0.574      -3.363       1.897
NW             4.2045      6.481      0.649      0.521      -9.013      17.422
U1         -5827.1027   4210.289     -1.384      0.176   -1.44e+04    2759.838
U2           167.7997     82.336      2.038      0.050      -0.126     335.725
Wealth         0.0962      0.104      0.928      0.361      -0.115       0.308
Ineq          70.6721     22.717      3.111      0.004      24.341     117.003
Prob       -4855.2658   2272.375     -2.137      0.041   -9489.804    -220.727
Time          -3.4790      7.165     -0.486      0.631     -18.093      11.135
==============================================================================
Omnibus:                        2.036   Durbin-Watson:                   1.723
Prob(Omnibus):                  0.361   Jarque-Bera (JB):                1.135
Skew:                           0.198   Prob(JB):                        0.567
Kurtosis:                       3.651   Cond. No.                     7.51e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 7.51e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [10]:
x = uc[['M', 'Ed', 'Po1', 'Po2', 'LF', 'M.F', 'Pop', 'NW', 'U1', 'U2',
       'Wealth', 'Ineq', 'Prob', 'Time']]
x = sm.add_constant(x)

In [11]:
model = sm.OLS(y,x)
result = model.fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Crime   R-squared:                       0.803
Model:                            OLS   Adj. R-squared:                  0.717
Method:                 Least Squares   F-statistic:                     9.322
Date:                Thu, 14 Oct 2021   Prob (F-statistic):           1.12e-07
Time:                        15:55:49   Log-Likelihood:                -308.02
No. Observations:                  47   AIC:                             646.0
Df Residuals:                      32   BIC:                             673.8
Df Model:                          14                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -5980.6789   1596.659     -3.746      0.001   -9232.967   -2728.391
M             87.7288     40.872      2.146      0.040       4.476     170.982
Ed           188.2299     61.003      3.086      0.004      63.971     312.489
Po1          192.7303    104.401      1.846      0.074     -19.927     405.388
Po2         -109.2166    115.358     -0.947      0.351    -344.193     125.760
LF          -646.0567   1274.654     -0.507      0.616   -3242.441    1950.328
M.F           17.3263     19.792      0.875      0.388     -22.989      57.642
Pop           -0.7331      1.269     -0.578      0.568      -3.319       1.852
NW             4.1347      5.786      0.715      0.480      -7.650      15.919
U1         -5786.2779   3834.471     -1.509      0.141   -1.36e+04    2024.284
U2           167.3329     79.023      2.118      0.042       6.367     328.298
Wealth         0.0956      0.099      0.962      0.343      -0.107       0.298
Ineq          70.4548     20.736      3.398      0.002      28.217     112.692
Prob       -4863.6294   2213.317     -2.197      0.035   -9372.008    -355.251
Time          -3.4549      6.991     -0.494      0.625     -17.696      10.786
==============================================================================
Omnibus:                        2.071   Durbin-Watson:                   1.724
Prob(Omnibus):                  0.355   Jarque-Bera (JB):                1.167
Skew:                           0.201   Prob(JB):                        0.558
Kurtosis:                       3.658   Cond. No.                     6.92e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 6.92e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [12]:
x = uc[['M', 'Ed', 'Po1', 'Po2', 'LF', 'M.F', 'Pop', 'NW', 'U1', 'U2',
       'Wealth', 'Ineq', 'Prob']]
x = sm.add_constant(x)

In [13]:
model = sm.OLS(y,x)
result = model.fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Crime   R-squared:                       0.802
Model:                            OLS   Adj. R-squared:                  0.723
Method:                 Least Squares   F-statistic:                     10.25
Date:                Thu, 14 Oct 2021   Prob (F-statistic):           3.76e-08
Time:                        15:56:22   Log-Likelihood:                -308.19
No. Observations:                  47   AIC:                             644.4
Df Residuals:                      33   BIC:                             670.3
Df Model:                          13                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -6285.4967   1455.742     -4.318      0.000   -9247.226   -3323.768
M             84.7505     39.959      2.121      0.042       3.453     166.048
Ed           189.5225     60.245      3.146      0.003      66.953     312.092
Po1          177.2062     98.416      1.801      0.081     -23.022     377.435
Po2          -89.4258    106.937     -0.836      0.409    -306.992     128.140
LF          -635.6554   1259.801     -0.505      0.617   -3198.739    1927.429
M.F           19.2693     19.174      1.005      0.322     -19.741      58.280
Pop           -0.8847      1.217     -0.727      0.473      -3.362       1.592
NW             3.3714      5.511      0.612      0.545      -7.841      14.584
U1         -5609.7329   3773.821     -1.486      0.147   -1.33e+04    2068.163
U2           164.2954     77.877      2.110      0.043       5.854     322.737
Wealth         0.0913      0.098      0.934      0.357      -0.108       0.290
Ineq          71.2477     20.436      3.486      0.001      29.671     112.824
Prob       -4212.7902   1758.346     -2.396      0.022   -7790.171    -635.409
==============================================================================
Omnibus:                        0.929   Durbin-Watson:                   1.701
Prob(Omnibus):                  0.628   Jarque-Bera (JB):                0.308
Skew:                           0.122   Prob(JB):                        0.857
Kurtosis:                       3.312   Cond. No.                     6.89e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 6.89e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [15]:
x = uc[['M', 'Ed', 'Po1', 'Po2', 'M.F', 'Pop', 'NW', 'U1', 'U2',
       'Wealth', 'Ineq', 'Prob']]
x = sm.add_constant(x)

In [16]:
model = sm.OLS(y,x)
result = model.fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Crime   R-squared:                       0.800
Model:                            OLS   Adj. R-squared:                  0.729
Method:                 Least Squares   F-statistic:                     11.34
Date:                Thu, 14 Oct 2021   Prob (F-statistic):           1.21e-08
Time:                        15:57:00   Log-Likelihood:                -308.37
No. Observations:                  47   AIC:                             642.7
Df Residuals:                      34   BIC:                             666.8
Df Model:                          12                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -6199.0822   1429.697     -4.336      0.000   -9104.577   -3293.587
M             89.3004     38.499      2.320      0.027      11.060     167.541
Ed           179.2903     56.104      3.196      0.003      65.273     293.307
Po1          165.5450     94.609      1.750      0.089     -26.724     357.814
Po2          -73.5251    101.062     -0.728      0.472    -278.908     131.858
M.F           14.3580     16.338      0.879      0.386     -18.845      47.561
Pop           -0.9847      1.188     -0.829      0.413      -3.399       1.430
NW             2.8819      5.365      0.537      0.595      -8.022      13.786
U1         -5012.5112   3543.900     -1.414      0.166   -1.22e+04    2189.560
U2           166.8588     76.854      2.171      0.037      10.672     323.045
Wealth         0.0899      0.097      0.930      0.359      -0.106       0.286
Ineq          71.1012     20.208      3.518      0.001      30.033     112.169
Prob       -4020.0278   1697.423     -2.368      0.024   -7469.607    -570.449
==============================================================================
Omnibus:                        0.615   Durbin-Watson:                   1.692
Prob(Omnibus):                  0.735   Jarque-Bera (JB):                0.156
Skew:                           0.108   Prob(JB):                        0.925
Kurtosis:                       3.183   Cond. No.                     6.50e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 6.5e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [18]:
x = uc[['M', 'Ed', 'Po1', 'Po2', 'M.F', 'Pop', 'U1', 'U2', 'Wealth', 'Ineq', 'Prob']]
x = sm.add_constant(x)

In [19]:
model = sm.OLS(y,x)
result = model.fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Crime   R-squared:                       0.798
Model:                            OLS   Adj. R-squared:                  0.735
Method:                 Least Squares   F-statistic:                     12.60
Date:                Thu, 14 Oct 2021   Prob (F-statistic):           3.75e-09
Time:                        15:57:33   Log-Likelihood:                -308.57
No. Observations:                  47   AIC:                             641.1
Df Residuals:                      35   BIC:                             663.3
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -6222.9676   1414.406     -4.400      0.000   -9094.365   -3351.570
M             97.5618     34.933      2.793      0.008      26.644     168.480
Ed           175.8207     55.162      3.187      0.003      63.837     287.805
Po1          161.2052     93.300      1.728      0.093     -28.205     350.615
Po2          -63.9087     98.448     -0.649      0.520    -263.768     135.951
M.F           13.1699     16.022      0.822      0.417     -19.357      45.697
Pop           -0.9468      1.174     -0.807      0.425      -3.330       1.436
U1         -5097.2535   3504.216     -1.455      0.155   -1.22e+04    2016.683
U2           170.7244     75.735      2.254      0.031      16.975     324.474
Wealth         0.0850      0.095      0.892      0.378      -0.108       0.278
Ineq          74.1141     19.216      3.857      0.000      35.104     113.124
Prob       -3743.4823   1600.931     -2.338      0.025   -6993.545    -493.420
==============================================================================
Omnibus:                        0.837   Durbin-Watson:                   1.606
Prob(Omnibus):                  0.658   Jarque-Bera (JB):                0.234
Skew:                           0.073   Prob(JB):                        0.890
Kurtosis:                       3.313   Cond. No.                     6.49e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 6.49e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [20]:
x = uc[['M', 'Ed', 'Po1', 'M.F', 'Pop', 'U1', 'U2', 'Wealth', 'Ineq', 'Prob']]
x = sm.add_constant(x)

In [21]:
model = sm.OLS(y,x)
result = model.fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Crime   R-squared:                       0.796
Model:                            OLS   Adj. R-squared:                  0.739
Method:                 Least Squares   F-statistic:                     14.04
Date:                Thu, 14 Oct 2021   Prob (F-statistic):           1.17e-09
Time:                        15:58:05   Log-Likelihood:                -308.85
No. Observations:                  47   AIC:                             639.7
Df Residuals:                      36   BIC:                             660.1
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -6403.9840   1375.459     -4.656      0.000   -9193.545   -3614.424
M             97.1997     34.647      2.805      0.008      26.933     167.467
Ed           171.8234     54.374      3.160      0.003      61.547     282.100
Po1          102.0204     19.662      5.189      0.000      62.145     141.896
M.F           15.1549     15.601      0.971      0.338     -16.485      46.795
Pop           -0.8776      1.159     -0.757      0.454      -3.229       1.474
U1         -5169.7981   3474.175     -1.488      0.145   -1.22e+04    1876.154
U2           172.4554     75.077      2.297      0.028      20.192     324.719
Wealth         0.0857      0.094      0.908      0.370      -0.106       0.277
Ineq          75.0703     19.005      3.950      0.000      36.527     113.614
Prob       -3758.9803   1587.837     -2.367      0.023   -6979.264    -538.697
==============================================================================
Omnibus:                        0.633   Durbin-Watson:                   1.594
Prob(Omnibus):                  0.729   Jarque-Bera (JB):                0.137
Skew:                          -0.072   Prob(JB):                        0.934
Kurtosis:                       3.222   Cond. No.                     6.49e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 6.49e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [22]:
x = uc[['M', 'Ed', 'Po1', 'M.F', 'Pop', 'U1', 'U2', 'Ineq', 'Prob']]
x = sm.add_constant(x)

In [23]:
model = sm.OLS(y,x)
result = model.fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Crime   R-squared:                       0.791
Model:                            OLS   Adj. R-squared:                  0.740
Method:                 Least Squares   F-statistic:                     15.58
Date:                Thu, 14 Oct 2021   Prob (F-statistic):           4.17e-10
Time:                        15:58:44   Log-Likelihood:                -309.39
No. Observations:                  47   AIC:                             638.8
Df Residuals:                      37   BIC:                             657.3
Df Model:                           9                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -6068.1757   1321.624     -4.591      0.000   -8746.041   -3390.311
M             91.1349     33.916      2.687      0.011      22.415     159.854
Ed           181.4908     53.194      3.412      0.002      73.709     289.273
Po1          108.7265     18.178      5.981      0.000      71.895     145.558
M.F           17.9369     15.260      1.175      0.247     -12.984      48.857
Pop           -0.7533      1.149     -0.656      0.516      -3.081       1.574
U1         -5855.3447   3383.024     -1.731      0.092   -1.27e+04     999.313
U2           187.5617     73.035      2.568      0.014      39.578     335.545
Ineq          64.2130     14.734      4.358      0.000      34.360      94.066
Prob       -4050.5933   1551.299     -2.611      0.013   -7193.823    -907.363
==============================================================================
Omnibus:                        1.644   Durbin-Watson:                   1.728
Prob(Omnibus):                  0.440   Jarque-Bera (JB):                0.810
Skew:                          -0.094   Prob(JB):                        0.667
Kurtosis:                       3.615   Cond. No.                     1.30e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.3e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [24]:
x = uc[['M', 'Ed', 'Po1', 'M.F', 'U1', 'U2', 'Ineq', 'Prob']]
x = sm.add_constant(x)

In [25]:
model = sm.OLS(y,x)
result = model.fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Crime   R-squared:                       0.789
Model:                            OLS   Adj. R-squared:                  0.744
Method:                 Least Squares   F-statistic:                     17.74
Date:                Thu, 14 Oct 2021   Prob (F-statistic):           1.16e-10
Time:                        15:59:13   Log-Likelihood:                -309.66
No. Observations:                  47   AIC:                             637.3
Df Residuals:                      38   BIC:                             654.0
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -6426.1010   1194.611     -5.379      0.000   -8844.464   -4007.738
M             93.3215     33.497      2.786      0.008      25.510     161.133
Ed           180.1201     52.753      3.414      0.002      73.327     286.914
Po1          102.6532     15.524      6.613      0.000      71.227     134.080
M.F           22.3397     13.601      1.642      0.109      -5.194      49.874
U1         -6086.6332   3339.265     -1.823      0.076   -1.28e+04     673.356
U2           187.3451     72.485      2.585      0.014      40.607     334.083
Ineq          61.3349     13.959      4.394      0.000      33.076      89.594
Prob       -3796.0318   1490.646     -2.547      0.015   -6813.688    -778.376
==============================================================================
Omnibus:                        0.815   Durbin-Watson:                   1.752
Prob(Omnibus):                  0.665   Jarque-Bera (JB):                0.238
Skew:                          -0.104   Prob(JB):                        0.888
Kurtosis:                       3.280   Cond. No.                     1.21e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.21e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [26]:
x = uc[['M', 'Ed', 'Po1', 'U1', 'U2', 'Ineq', 'Prob']]
x = sm.add_constant(x)

In [27]:
model = sm.OLS(y,x)
result = model.fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Crime   R-squared:                       0.774
Model:                            OLS   Adj. R-squared:                  0.733
Method:                 Least Squares   F-statistic:                     19.06
Date:                Thu, 14 Oct 2021   Prob (F-statistic):           8.81e-11
Time:                        15:59:41   Log-Likelihood:                -311.27
No. Observations:                  47   AIC:                             638.5
Df Residuals:                      39   BIC:                             653.3
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -5095.5521    896.895     -5.681      0.000   -6909.694   -3281.410
M            106.7818     33.179      3.218      0.003      39.672     173.892
Ed           218.4468     48.331      4.520      0.000     120.689     316.205
Po1          105.9552     15.725      6.738      0.000      74.149     137.761
U1         -3542.3454   3021.939     -1.172      0.248   -9654.795    2570.104
U2           158.8153     71.888      2.209      0.033      13.408     304.223
Ineq          66.3349     13.917      4.767      0.000      38.186      94.484
Prob       -3730.8489   1522.207     -2.451      0.019   -6809.803    -651.895
==============================================================================
Omnibus:                        3.383   Durbin-Watson:                   1.923
Prob(Omnibus):                  0.184   Jarque-Bera (JB):                2.675
Skew:                          -0.222   Prob(JB):                        0.262
Kurtosis:                       4.081   Cond. No.                     2.88e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.88e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [28]:
x = uc[['M', 'Ed', 'Po1', 'U2', 'Ineq', 'Prob']]
x = sm.add_constant(x)

In [29]:
model = sm.OLS(y,x)
result = model.fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Crime   R-squared:                       0.766
Model:                            OLS   Adj. R-squared:                  0.731
Method:                 Least Squares   F-statistic:                     21.81
Date:                Thu, 14 Oct 2021   Prob (F-statistic):           3.42e-11
Time:                        16:00:05   Log-Likelihood:                -312.08
No. Observations:                  47   AIC:                             638.2
Df Residuals:                      40   BIC:                             651.1
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -5040.5050    899.843     -5.602      0.000   -6859.156   -3221.854
M            105.0196     33.299      3.154      0.003      37.719     172.320
Ed           196.4712     44.754      4.390      0.000     106.019     286.923
Po1          115.0242     13.754      8.363      0.000      87.227     142.821
U2            89.3660     40.906      2.185      0.035       6.693     172.039
Ineq          67.6532     13.936      4.855      0.000      39.488      95.818
Prob       -3801.8363   1528.097     -2.488      0.017   -6890.236    -713.436
==============================================================================
Omnibus:                        2.955   Durbin-Watson:                   1.932
Prob(Omnibus):                  0.228   Jarque-Bera (JB):                2.360
Skew:                          -0.001   Prob(JB):                        0.307
Kurtosis:                       4.098   Cond. No.                     1.45e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.45e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

R^2 is 0.766, the coefficients are: M = 105.0196, Ed = 196.4712, Po1 = 115.0242, U2 = 89.3660, Ineq = 67.6532, and Prob = -3801.8363. All of the variables have p-value of 0.0 or close to 0.0.

# Problem 2  [15]
Filter the uc data with only the columns you found in the final model of problem 1 and the Crime column. Now split the data to testing and training following 75%, 25% rule. Use only the training data to fit the linear model.  Report the $R^2$ and adjusted $R^2$ for the training data. How do these compare with the $R^2$ and adjusted $R^2$ of the final model from problem 1? Explain why there is a difference in these numbers if any.

In [36]:
from sklearn.model_selection import train_test_split

In [56]:
xx = uc[['M', 'Ed', 'Po1', 'U2', 'Ineq', 'Prob']]
y = uc.Crime
xtrain, xtest, ytrain, ytest = model_selection.train_test_split(xx, y, test_size = 0.25, random_state = 17)

In [57]:
modelxx = sm.OLS(ytrain, sm.add_constant(xtrain)).fit()
modelxx.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Crime   R-squared:                       0.782
Model:                            OLS   Adj. R-squared:                  0.736
Method:                 Least Squares   F-statistic:                     16.77
Date:                Fri, 15 Oct 2021   Prob (F-statistic):           4.10e-08
Time:                        15:15:26   Log-Likelihood:                -233.94
No. Observations:                  35   AIC:                             481.9
Df Residuals:                      28   BIC:                             492.8
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -5615.9991   1048.444     -5.357      0.000   -7763.640   -3468.358
M             77.1977     45.302      1.704      0.099     -15.600     169.995
Ed           235.2465     56.588      4.157      0.000     119.330     351.163
Po1          122.2868     16.755      7.298      0.000      87.966     156.608
U2            95.4822     49.017      1.948      0.062      -4.925     195.890
Ineq          89.1067     19.416      4.589      0.000      49.334     128.879
Prob       -2878.7053   1995.060     -1.443      0.160   -6965.400    1207.989
==============================================================================
Omnibus:                        3.592   Durbin-Watson:                   1.544
Prob(Omnibus):                  0.166   Jarque-Bera (JB):                2.666
Skew:                           0.220   Prob(JB):                        0.264
Kurtosis:                       4.278   Cond. No.                     1.53e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.53e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

The correlations of the training set R^2 (0.782) and the final model R^2 (0.766) are almost the same. This means that the relationship of the variables in the training set and the final model are very similar.
The adjusted R^2 of the training set is 0.736 and in the final model is (0.731) are also almost the same. The relationship of the variables is similar.

# Problem 3 [15]
Use the model from problem 2 to predict crime values using the testing data. Make a table of $R^2$. Adjusted-$R^2$, Root mean squared error (RMSE) and Mean Absolute Error (MAE) for the training and testing data sets. Compare and comment on all the numbers for training and testing data set.  Which of the 4 different measures is best for evaluating performance with testing and training?  Explain your reasoning.

In [58]:
ypred = modelxx.predict(sm.add_constant(xtest))
ypred

46     990.269958
40     678.844983
16     476.449590
29     611.762732
21     790.782223
38     785.548024
41     333.651671
37     588.215765
43    1158.018446
33    1014.693705
35     933.659881
11     615.075454
dtype: float64

In [66]:
from sklearn.metrics import r2_score
r2 = r2_score(ytest, ypred)
r2

0.3103042882131918

In [67]:
adj = 1 - (1-r2_score(ytest, ypred)) * (len(y)-1)/(len(y)-x.shape[1]-1)
adj

0.186512750200175

In [60]:
from sklearn.metrics import mean_squared_error
from math import sqrt

In [61]:
rmse = sqrt(mean_squared_error(ytest, ypred))
rmse

190.33082207861545

In [62]:
from sklearn.metrics import mean_absolute_error
mean_absolute_error(ytest, ypred)

158.66564688797703

In [68]:
ytrainpred = modelxx.predict(sm.add_constant(xtrain))
ytrainpred

8      698.694184
24     634.644054
28    1498.369723
18    1154.301341
12     793.631566
2      426.865530
19    1204.261895
30     368.299403
23     927.542586
0      873.311455
26     178.861686
5      649.977343
13     734.414827
14     865.158114
36     906.857858
25    1824.225198
9      760.431497
42     976.441191
32     898.628737
27    1268.094049
45     727.784283
3     1917.422087
34     744.937073
10    1093.375114
17     826.881223
1     1392.967248
7     1420.128981
20     789.314530
39    1168.614659
4     1274.143210
31     753.838453
44     652.761917
22     965.161770
6      717.031588
15    1040.625628
dtype: float64

In [70]:
r2 = r2_score(ytrain, ytrainpred)
r2

0.7822672306364259

In [71]:
adj = 1 - (1-r2_score(ytrain, ytrainpred)) * (len(y)-1)/(len(y)-x.shape[1]-1)
adj

0.7431869899814254

In [72]:
rmse = sqrt(mean_squared_error(ytrain, ytrainpred))
rmse

193.44534410599067

In [73]:
mean_absolute_error(ytrain, ytrainpred)

143.99068413562372

After finding values, you can click the following table and just fill in the values.

|Measure | Training Data | Testing Data|
|:-|:-:|-:|
|**$R^2$**|$0.782$|$0.310$|
|**Adjusted $R^2$**|$0.743$|$0.187$|
|**RMSE**|$193.445$|$190.331$|
|**MAE**|$143.991$|$158.666$|



The R^2 and the adjusted R^2 values are more highly positively correlated in the training data set than the testing data set. The errors for both sets are closely similar. R^2 is the best because higher value means more accurate representation of the data. It measures the difference between the observed data and the values, so higher r^2 means smaller margin of error.